lib/loader.js

```diff
@@ -0,0 +1,5 @@
+'use strict';
+
+const Loader = require('internal/loader/Loader');
+
+module.exports = new Loader();
```

Comment:
I am not comfortable exposing the loader when hooks are not even finished. Please remove this.

Reply:
if its marked as experimental is that ok?

Reply:
I'd still lean on no. I'm not entirely sure why you want it exposed if you want people to use hooks to mutate it.

Reply:
@bmeck Node’s lack of APIs for its internals has always been a *huge* pain point. We should definitely expose the `Loader` at some point.

> if you want people to use hooks to mutate it.

That makes sense if you think of the loader as Node’s way of loading its users’ modules from disk. Node definitely *should* have some kind of API for people to create and build modules. Think of it like the `vm` module, just more complicated because modules are more complicated than scripts.

Reply:
Btw, this is going to conflict with @JacksonTian’s npm.im/loader

Reply:
> That makes sense if you think of the loader as Node’s way of loading its users’ modules from disk. Node definitely should have some kind of API for people to create and build modules. Think of it like the vm module, just more complicated because modules are more complicated than scripts.

@addaleax that is why I have been doing things like https://github.com/w3c/FileAPI/issues/84 and watching https://github.com/TypeStrong/ts-node/issues/436 which wants in-memory creation of things.

We don't need to create a `vm` API for this with that and/or the TC39 Realms proposal.

Reply:
I should note that there might be *multiple loaders* so thats also something to think about. Having multiple allows composability (even if just by convention). This is especially likely w/ per-package hooks which are desired for signed things like https://github.com/WICG/webpackage to get loading guarantees. I would think some sore of loader registry might be more desirable than a single mutable root. Registration would also have some benefits like detecting if a loader is already active so people don't run code coverage/transpilation multiple times accidentally.

Reply:
@bmeck That still seems only concerned loading modules in the context of Node’s module loader?

Basically, I do think we need to expose a way to create indepenent in-memory loaders that can be hooked without disk accesses.

`--loader` is perfectly fine but has a different use case

Reply:
@bmeck Oh, right, this should expose the `Loader` class, not (only) the main instance. My bad.

Reply:
@addaleax +1, I agree we will need to expose some sort of API to make your own loader. Not convinced we should ever expose the main instance though.

Reply:
> we will need to expose some sort of API to make your own loader.

i meant to expose that, accidentally deleted a line :3

> Not convinced we should ever expose the main instance though.

I would like to expose this to allow users to have some control (and fun?) with node. I will agree with addalex that the locked down state of node is definitely a painful thing.

Reply:
> I would like to expose this to allow users to have some control (and fun?) with node. I will agree with addalex that the locked down state of node is definitely a painful thing.

We should then define the exact thing we are exposing. Having it be dynamically mutable is a big API to expose. This would mean the behavior of the loader can change over the lifetime of a process which doesn't have a clear use case to me. It would also mean that people might need to lock down the loader object if they are doing defensive programming. Having it be declarative using things like `package.json` per-package hooking ([some talking about this in a gist](https://gist.github.com/bmeck/5b43052cb2e28bdae29c161f16f12801)) would have a few advantages.

* Can be picked up by non-node tooling such as bundlers.
* Standardized place to look at for where hooks are instantiated.

Minimal exposure to userland would probably need 2 different instances exposed I think. We expose the loader as it exists after `--loader` or should we expose the "default" loader pre `--loader` as well.

Is there a reason to expose it to runtime? What control are you seeking to gain that you cannot achieve right now and/or is painful to achieve?

Reply:
how about exposing `Loader` by default so people can make their own, and only exposing the main instance with `--loader`

Reply:
@devsnek I'd need to do a code review of it since I didn't design it for public exposure.

Reply:
@devsnek is there a use case you are trying to get by doing this? I keep asking the question but don't understand right now.

Reply:
I would like to expose it to allow for more control over node's loader, which will be a very important system moving forward. In the same way that require can be messed with, I think that loader should be able to be messed with. I agree that there should be control over what stuff can be messed with, but the loader having a public api is, in my opinion, a must have moving forward. I think that the hooks idea is good as a controlled system (and should remain in), but that it should also be possible to make your own loaders, change them up, etc. More to this point is why i suggested that at least exposing the class by default while allowing access to the instance with a flag would be a good choice to keep the system save and extensible at the same time.

Reply:
I would need to understand various aspects of this. I see you want it, but don't see why except an opinion of "I think that loader should be able to be messed with". Various parts of mutating require like `NODE_PATH` and `require.extensions` were deprecated and I don't feel that saying `require` is mutable so the future should be mutable is a strong position. Exposing individual use cases like hooking into how specifier resolution works is very different from creating multiple disconnected loaders.

In particular, this exposure is problematic since the Module Map is not shared. I would not expose this without a clearer reasoning to need/want multiple completely disconnected loaders.

I also feel like the [Realms API](https://github.com/tc39/proposal-realms) should be looked at to ensure we don't have problems if/when it arrives. The people who worked on the [WHATWG Loader API](https://github.com/whatwg/loader) and involved in that. It doesn't look to be exposing a Loader like object from the last few TC39 meetings on this topic.

If you can explain the concrete use case of having multiple loaders and explain how it does not suffer from the problems of mutation in the same way that `require` has that seems fine. Right now this seems too rushed and hasn't looked at compatibility or scope of what exposing it means.

Reply:
"Messing with" the loader won't just have effects for you and your app, it could very very easily leak out and have effects on the ecosystem.

Whether this is a good idea *ever* is a separate discussion, but allowing this kind of experimentation prior to everything being stable and finished is, imo, a very dangerous and bad idea.